### PR TITLE
Implement $set stage dsl

### LINKED ISSUE #72: Implement $set stage dsl
https://www.mongodb.com/docs/manual/reference/operator/aggregation/set/

### PR DESCRIPTION
**Pull request by username1103**
resolve #72

### CODE REVIEW

File: example/spring-data-mongodb/src/main/kotlin/com/github/inflab/example/spring/data/mongodb/repository/SetRepository.kt [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+        val aggregation = aggregation {
+            match(Criteria.where(Score::id.toDotPath()).isEqualTo(1))
+            set {
+                // TODO: add $concatArrays expression
+                Score::homework set Document("\$concatArrays", listOf("\$${Score::homework.toDotPath()}", listOf(7)))
+            }
+        }
+
+        return mongoTemplate.aggregate<Score>(aggregation, SCORES)
+    }
```
Comment (username1103) [inline example/spring-data-mongodb/src/main/kotlin/com/github/inflab/example/spring/data/mongodb/repository/SetRepository.kt:131]: In example, we have to add `7` at the end of "homework" array.
```
db.scores.aggregate([
   { $match: { _id: 1 } },
   { $set: { homework: { $concatArrays: [ "$homework", [ 7 ] ] } } }
])
```

But, ConcatArrays not supported this. We can only concat value at the beginning maybe..

https://github.com/spring-projects/spring-data-mongodb/blob/69285151a0a8fc30b90cda06505bef973ea7151b/spring-data-mongodb/src/test/java/org/springframework/data/mongodb/core/aggregation/ArrayOperatorsUnitTests.java#L77-L82

File: example/spring-data-mongodb/src/main/kotlin/com/github/inflab/example/spring/data/mongodb/repository/UnionWithRepository.kt [hunk 2]
```diff
@@ -24,33 +23,40 @@ class UnionWithRepository(
      * @see <a href="https://www.mongodb.com/docs/manual/reference/operator/aggregation/unionWith/#report-1--all-sales-by-year-and-stores-and-items">All Sales by Year and Stores and Items</a>
      */
     fun findSalesByYearAndStoresAndItems(): AggregationResults<SalesDto> {
-        // TODO: apply $set dsl
         val aggregation = aggregation {
-            stage(SetOperation.set("_id").toValue("2017"))
+            set {
+                SalesDto::id set "2017"
+            }
```
Comment (jbl428) [inline example/spring-data-mongodb/src/main/kotlin/com/github/inflab/example/spring/data/mongodb/repository/UnionWithRepository.kt:29]: 👍 👍 👍 👍

File: example/spring-data-mongodb/src/main/kotlin/com/github/inflab/example/spring/data/mongodb/repository/SetRepository.kt (hunk 1 above)
Comment (jbl428) [reply example/spring-data-mongodb/src/main/kotlin/com/github/inflab/example/spring/data/mongodb/repository/SetRepository.kt:131]: how about using `literal` expression. `literals(listof(7))`
but I can not sure if it works

Comment (username1103) [reply example/spring-data-mongodb/src/main/kotlin/com/github/inflab/example/spring/data/mongodb/repository/SetRepository.kt]: 👍🏻👍🏻👍🏻👍🏻👍🏻👍🏻👍🏻👍🏻👍🏻👍🏻👍🏻👍🏻

Comment (jbl428) [reply example/spring-data-mongodb/src/main/kotlin/com/github/inflab/example/spring/data/mongodb/repository/SetRepository.kt]: Hmm, I think it is better to implement `$concatArrays` without using ArrayOperator from spring.
It seems fine to leave it as it for now :)

**Review by jbl428 — APPROVED** [review]
LGTM

**Pull request merged**